Document behavior of endpoints with the feature EndpointSliceTerminatingCondition (#36791)
kubernetes/website: content/en/docs/tutorials/services/pods-and-endpoint-termination-flow.md

---
title: Pods And Endpoints Termination Flow
content_type: tutorial
weight: 60
---

<!-- overview -->

## The Kubernetes model for connecting containers

> Now that you have a continuously running, replicated application, that is exposed on a network, let's look at Pods termination flow and ways to implement a graceful connections draining.

Reviewer: In a new tutorial page like this, we may want to polish the first sentence.

Also, please wrap long lines when appropriate.

Author: I wrapped lines. Is there any suggestions on how to make text better?

There are often a need to terminate a Pod - be it for upgrade or scale down. In order to improve application availability, it may be important to implement a proper active connections draining. This tutorial explains the flow of Pod termination in connection with the corresponding endpoint state and removal.

This tutorial uses a simple nginx web server to demonstrate the concept.

<!-- body -->

#### Example flow with endpoint termination

> The following is the example of the flow described in the [Termination of Pods](/concepts/workloads/pods/pod-lifecycle/#pod-termination) document.

Reviewer: Note the link has a `/docs` prefix.

```suggestion
The following is the example of the flow described in the
[Termination of Pods](/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination) document.
```

        lifecycle:
          preStop:
            exec:
              # just hang around for the duration of terminationGracePeriodSeconds
              command: [
                "/bin/sh", "-c", "sleep 120"
              ]

---

apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  selector:
    app: nginx
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
```

Once the Pod and Service are running, you can get the name of EndpointSlice:

```shell
kubectl get endpointslice
```

The output is similar to this:

```shell
NAME                  ADDRESSTYPE   PORTS   ENDPOINTS                 AGE
nginx-service-6tjbr   IPv4          80      10.12.1.199,10.12.1.201   22m
```

You can see it's status and validate that there is one Endpoint registered:

```shell
kubectl get endpointslices -o json -l kubernetes.io/service-name=nginx-service
```

The output is similar to this:

```json
{
    "addressType": "IPv4",
    "apiVersion": "discovery.k8s.io/v1",
    "endpoints": [
        {
            "addresses": [
                "10.12.1.201"
            ],
            "conditions": {
                "ready": true,
                "serving": true,

>                 "terminating": false

Reviewer: This snippet will be syntax highlighted as a JSON object.
Since the object is incomplete, e.g. no ending brackets etc, the syntax parsing may fail.

If we only want to show an incomplete JSON object, we can replace `json` with `none` on line 94.

Author: most syntax highlighter ok with partial json. But happy to change, I understand the desire to make it as portable as possible.

```

Now let's terminate the Pod and validate that the Pod is being terminated
respecting the graceful termination period configuration:

```shell
kubectl delete pod nginx-deployment-7768647bf9-b4b9s
```

All pods:

```shell
kubectl get pods
```

The output is similar to this:

```shell
NAME                                READY   STATUS        RESTARTS      AGE
nginx-deployment-7768647bf9-b4b9s   1/1     Terminating   0             4m1s
nginx-deployment-7768647bf9-rkxlw   1/1     Running       0             8s
```

You can see that the new pod got scheduled.

With the feature gate `EndpointSliceTerminatingCondition` enabled by default
while the new Endpoint is being created for the new Pod, the old endpoint is
still around in the terminating state:

```shell
kubectl get endpointslice -o json nginx-service-6tjbr
```

The output is similar to this:

> ```json

Reviewer: Same as above

---
title: Pods And Endpoints Termination Flow
content_type: tutorial
weight: 60
---

<!-- overview -->

> ## The Kubernetes model for connecting containers

Reviewer 2: Avoid starting a page with a heading. Either omit the heading, or put some text before it.

> Now that you have a continuously running, replicated application, that is

Reviewer: > Now that you have ...

Nothing has been discussed before this sentence or this paragraph, because this is the very first sentence in the page. The expression "Now that you have ..." reads pretty confusing to me.

Reviewer 2: This wording comes from https://kubernetes.io/docs/tutorials/services/connect-applications-service/ (and needs fixing there too).

exposed on a network, let's look at Pods termination flow and ways to implement
a graceful connections draining.

> There are often a need to terminate a Pod - be it for upgrade or scale down.

Reviewer 3: Grammatical nit. Also we should also be explicit about what is being upgraded/scaled down.
```suggestion
There are often cases when you need to terminate a Pod - be it for upgrade or scale down.
```

In order to improve application availability, it may be important to implement
a proper active connections draining. This tutorial explains the flow of
Pod termination in connection with the corresponding endpoint state and removal.

> This tutorial uses a simple nginx web server to demonstrate the concept.

Reviewer 3: Can we not append this to the previous sentence as - This tutorial explains the flow of
Pod termination in connection with the corresponding endpoint state and removal by using a simple nginx web server to demonstrate the concept?

<!-- body -->

> #### Example flow with endpoint termination

Reviewer 2: ```suggestion
## Example flow with endpoint termination
```

The following is the example of the flow described in the
[Termination of Pods](/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)
document.

Let's say you have a Deployment containing of a single `nginx` replica
(just for demonstration purposes) and a Service:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      terminationGracePeriodSeconds: 120
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
        lifecycle:
          preStop:
            exec:
              # just hang around for the duration of terminationGracePeriodSeconds
              command: [
                "/bin/sh", "-c", "sleep 120"
              ]

---

apiVersion: v1
kind: Service
metadata:
  name: nginx-service
spec:
  selector:
    app: nginx
  ports:
    - protocol: TCP
      port: 80
      targetPort: 80
```

> Once the Pod and Service are running, you can get the name of EndpointSlice:

Reviewer 2: ```suggestion
Once the Pod and Service are running, you can get the name of any associated EndpointSlices:
```

```shell
kubectl get endpointslice
```

The output is similar to this:

```shell
NAME                  ADDRESSTYPE   PORTS   ENDPOINTS                 AGE
nginx-service-6tjbr   IPv4          80      10.12.1.199,10.12.1.201   22m
```

> You can see it's status and validate that there is one Endpoint registered:

Reviewer 2: ```suggestion
You can see its status, and validate that there is one endpoint registered:
```

```shell
kubectl get endpointslices -o json -l kubernetes.io/service-name=nginx-service
```

The output is similar to this:

```none
{
    "addressType": "IPv4",
    "apiVersion": "discovery.k8s.io/v1",
    "endpoints": [
        {
            "addresses": [
                "10.12.1.201"
            ],
            "conditions": {
                "ready": true,
                "serving": true,
                "terminating": false
```

Now let's terminate the Pod and validate that the Pod is being terminated
respecting the graceful termination period configuration:

```shell
kubectl delete pod nginx-deployment-7768647bf9-b4b9s
```

All pods:

```shell
kubectl get pods
```

The output is similar to this:

```shell
NAME                                READY   STATUS        RESTARTS      AGE
nginx-deployment-7768647bf9-b4b9s   1/1     Terminating   0             4m1s
nginx-deployment-7768647bf9-rkxlw   1/1     Running       0             8s
```

You can see that the new pod got scheduled.

With the feature gate `EndpointSliceTerminatingCondition` enabled by default

> while the new Endpoint is being created for the new Pod, the old endpoint is

Reviewer 2: ```suggestion
while the new endpoint is being created for the new Pod, the old endpoint is
```

    "apiVersion": "discovery.k8s.io/v1",
    "endpoints": [
        {
            "addresses": [
                "10.12.1.201"
            ],
            "conditions": {
                "ready": false,
                "serving": true,
                "terminating": true
            },
            "nodeName": "gke-main-default-pool-dca1511c-d17b",
            "targetRef": {
                "kind": "Pod",
                "name": "nginx-deployment-7768647bf9-b4b9s",
                "namespace": "default",
                "uid": "66fa831c-7eb2-407f-bd2c-f96dfe841478"
            },
            "zone": "us-central1-c"
        },
        {
            "addresses": [
                "10.12.1.202"
            ],
            "conditions": {
                "ready": true,
                "serving": true,
                "terminating": false
            },
            "nodeName": "gke-main-default-pool-dca1511c-d17b",
            "targetRef": {
                "kind": "Pod",
                "name": "nginx-deployment-7768647bf9-rkxlw",
                "namespace": "default",
                "uid": "722b1cbe-dcd7-4ed4-8928-4a4d0e2bbe35"
            },
            "zone": "us-central1-c"
```

> This allows to implement a connections draining functionality - clients

Reviewer 3: Who does it allow? You/users/clients?

Author: fixed

(like load balancers) may detect terminating endpoints and implement
a special logic for them.

> Terminating Endpoints always have `ready` status as `false` for backward

Reviewer 2: ```suggestion
In Kubernetes, endpoints that are terminating always have there `ready` status set as as `false`.
This needs to happen for backward
```

# Pods And Endpoints Termination Flow

There are often cases when you need to terminate a Pod - be it for upgrade or scale down.
In order to improve application availability, it may be important to implement
a proper active connections draining. This tutorial explains the flow of
Pod termination in connection with the corresponding endpoint state and removal.

This tutorial explains the flow of Pod termination in connection with the
corresponding endpoint state and removal by using
a simple nginx web server to demonstrate the concept.

<!-- body -->

## Example flow with endpoint termination

The following is the example of the flow described in the
[Termination of Pods](/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination)
document.

Let's say you have a Deployment containing of a single `nginx` replica
(just for demonstration purposes) and a Service:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:

>       terminationGracePeriodSeconds: 120

Reviewer 2: ```suggestion
      terminationGracePeriodSeconds: 120 # extra long grace period
```

      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
        lifecycle:
          preStop:
            exec:

>               # just hang around for the duration of terminationGracePeriodSeconds
>               command: [
>                 "/bin/sh", "-c", "sleep 120"

Reviewer 2: ```suggestion
              # just hang around for at least the duration of terminationGracePeriodSeconds
              command: [
                "/bin/sh", "-c", "sleep 180"
```

to avoid people inferring that the two numbers somehow need to match

---

or, if they _do_ need to match: explain why.

>               ]

Author: done

```shell
kubectl get endpointslices -o json -l kubernetes.io/service-name=nginx-service
```

The output is similar to this:

```none
{
    "addressType": "IPv4",
    "apiVersion": "discovery.k8s.io/v1",
    "endpoints": [
        {
            "addresses": [
                "10.12.1.201"
            ],
            "conditions": {
                "ready": true,
                "serving": true,
                "terminating": false
```

Now let's terminate the Pod and validate that the Pod is being terminated
respecting the graceful termination period configuration:

```shell
kubectl delete pod nginx-deployment-7768647bf9-b4b9s
```

All pods:

```shell
kubectl get pods
```

The output is similar to this:

```shell
NAME                                READY   STATUS        RESTARTS      AGE
nginx-deployment-7768647bf9-b4b9s   1/1     Terminating   0             4m1s
nginx-deployment-7768647bf9-rkxlw   1/1     Running       0             8s
```

You can see that the new pod got scheduled.

With the feature gate `EndpointSliceTerminatingCondition` enabled by default
while the new endpoint is being created for the new Pod, the old endpoint is

> still around in the terminating state:

Reviewer 2: Usually we tell readers that the doc is only relevant if they're running Kubernetes t least as new as [whatever]; we rarely explain things more finely than that.

People running a different version are welcome to go and read the docs for that other version, if it's in support.
What we should explain then is that this doc assumes you're running v1.26, and then we don't need to mention the feature gate at all.

> still around in the terminating state:

Author: changed the assumption - now text is written with the assumption that the feature gate does not exist any longer

```shell

> kubectl get endpointslice -o json nginx-service-6tjbr

Author: I think simple commands is something people would prefer to run. Documentation with advanced querying of specific fields is not what people would run day by day.

Also for readability, previous advice making it `none` makes json harder to read. If we can trust that parsers would be OK with incomplete json, syntax highlighting will make it easy to understand where to look at. I can also close brackets for the parser sake
